llm-chain-local/src/executor.rs

```diff
@@ -0,0 +1,195 @@
+use std::path::Path;
+
+use crate::options::{PerExecutor, PerInvocation};
+use crate::output::Output;
+use crate::LocalLlmTextSplitter;
+
+use async_trait::async_trait;
+use llm_chain::prompt::Prompt;
+use llm_chain::tokens::{PromptTokensError, Tokenizer, TokenizerError};
+use llm_chain::traits::ExecutorCreationError;
+
+/// Executor is responsible for running the LLM and managing its context.
+pub struct Executor {
+    llm: Box<dyn llm::Model>,
+    callback: Option<fn(&Output)>,
+}
+
+impl Executor {
+    pub fn with_callback(mut self, callback: fn(&Output)) -> Self {
+        self.callback = Some(callback);
+        self
+    }
+
+    pub(crate) fn get_llm(&self) -> &dyn llm::Model {
+        self.llm.as_ref()
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("unable to tokenize prompt")]
+    PromptTokensError(PromptTokensError),
+    #[error("unable to create executor: {0}")]
+    InnerError(#[from] Box<dyn std::error::Error>),
+}
+
+impl llm_chain::traits::ExecutorError for Error {}
+
+#[async_trait]
+impl llm_chain::traits::Executor for Executor {
+    type PerInvocationOptions = PerInvocation;
+    type PerExecutorOptions = PerExecutor;
+    type Output = Output;
+    type Error = Error;
+    type Token = i32;
+    type StepTokenizer<'a> = LocalLlmTokenizer<'a>;
+    type TextSplitter<'a> = LocalLlmTextSplitter<'a>;
+
+    fn new_with_options(
+        options: Option<Self::PerExecutorOptions>,
+        invocation_options: Option<Self::PerInvocationOptions>,
+    ) -> Result<Self, llm_chain::traits::ExecutorCreationError> {
+        let model_type = options
+            .as_ref()
+            .and_then(|x| x.model_type.clone())
+            .or_else(|| std::env::var("LLM_MODEL_TYPE").ok())
+            .ok_or(ExecutorCreationError::FieldRequiredError(
+                "model_type, ensure to provide the parameter or set `LLM_MODEL_TYPE` environment variable ".to_string(),
+            ))?;
+        let model_path = options
+                .as_ref()
+                .and_then(|x| x.model_path.clone())
+                .or_else(|| std::env::var("LLM_MODEL_PATH").ok())
+                .ok_or(ExecutorCreationError::FieldRequiredError(
+                    "model_path, ensure to provide the parameter or set `LLM_MODEL_PATH` environment variable ".to_string(),
+                ))?;
+
+        let params = invocation_options.unwrap_or_default().into();
+        let llm: Box<dyn llm::Model> = match model_type.as_str() {
```

Comment:
Could this be done upstream?

Reply:
So you mean exposing a `load` option that takes an enum variant referring to the model type?

Reply:
meant the mapping of strings to model types :)